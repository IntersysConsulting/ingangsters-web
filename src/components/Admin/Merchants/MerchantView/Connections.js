import { API } from "../../../../config";
import axios from "axios";
import store from "../../../../store";
import {
  createNotificationSuccess,
  createNotificationError,
  createNotificationInfo
} from "../../../../actions/creators/notification";
import { idNotificationGenerator } from "../../../../utils/idGenerator";

export async function loadAdmin(id, setData) {
  const config = {
    headers: {
      Authorization: "Bearer " + localStorage.getItem("token"),
      "Content-Type": "application/json"
    }
  };
  const endpoint = `${API}/admin/get`;
  const body = {
    _id: id
  };
  try {
    const res = await axios.post(endpoint, body, config);
    if (res.status === 200) {
      const fullName = res.data.data.name.split(" ");
      const firstName = fullName.shift();
      setData(prev => ({
        ...prev,
        showDeleteButton: true,
        title: "Update Administrator",
        name: firstName,
        lastName: fullName.join(" "),
        email: res.data.data.email,
        submitAction: evt => {
          updateAdmin(evt, id);
        },
        isLoading: false,
        requirePassword: false
      }));
    }
  } catch (err) {
    window.location.replace("/404");
  }
}

export async function deleteAdmin(id) {
  const endpoint = `${API}/admin/delete`;
  const config = {
    headers: {
      "Content-Type": "application/json",
      Authorization: "Bearer " + localStorage.getItem("token")
    },
    data: {
      _id: id
    }
  };
  try {
    const res = await axios.delete(endpoint, config);
    if (res.status === 200) {
      createNotificationSuccess(
        idNotificationGenerator(),
        "Admin deleted",
        "Saved successfully"
      )(store.dispatch);
      setTimeout((window.location.replace("/admin/dashboard"), 5000));
    }
  } catch (err) {
    createNotificationError(
      idNotificationGenerator(),
      "Something went wrong",
      "The user wasn't deleted, try again later"
    )(store.dispatch);
  }
}

export async function createNewAdmin(evt) {
  const name = [
      evt.target["merchantName"].value,
      evt.target["merchantLastName"].value
    ].join(" "),
    email = evt.target["merchantEmail"].value,
    password1 = evt.target["merchantPassword"].value,
    password2 = evt.target["merchantPasswordConfirm"].value;

  if (password1 !== password2) {
    createNotificationInfo(
      idNotificationGenerator(),
      "Passwords must match",
      "Please check the fields"
    );
    evt.target["merchantPasswordConfirm"].focus();
  } else {
    const endpoint = `${API}/admin/create`;
    const config = {
      headers: {
        Authorization: "Bearer " + localStorage.getItem("token"),
        "Content-Type": "application/json"
      }
    };
    const data = {
      name: name,
      email: email,
      password: password1
    };
    try {
      const res = await axios.post(endpoint, data, config);
      if (res.status === 200) {
        createNotificationSuccess(
          idNotificationGenerator(),
          "Admin created",
          "Admin user created successfully"
        )(store.dispatch);
        setTimeout((window.location.pathname = "admin/dashboard"), 5000);
      } else {
        createNotificationError(
          idNotificationGenerator(),
          "Something went wrong",
          "Admin user was not created, try again later"
        )(store.dispatch);
        setTimeout(window.location.reload(), 5000);
      }
    } catch (err) {
      createNotificationError(
        idNotificationGenerator(),
        "Something went wrong",
        "Try again later"
      )(store.dispatch);
    }
  }
}

export async function updateAdmin(evt, id) {
  const password1 = evt.target["merchantPassword"].value,
    password2 = evt.target["merchantPasswordConfirm"].value;

  if (password1 !== password2) {
    createNotificationInfo(
      idNotificationGenerator(),
      "Passwords must match",
      "Please check the fields"
    );
    evt.target["merchantPasswordConfirm"].focus();
  } else {
    console.log("Updating...");
    const name = [
        evt.target["merchantName"].value,
        evt.target["merchantLastName"].value
      ].join(" "),
      email = evt.target["merchantEmail"].value,
      endpoint = `${API}/admin/update`,
      config = {
        headers: {
          Authorization: "Bearer " + localStorage.getItem("token"),
          "Content-Type": "application/json"
        }
      };
    const data = {
      _id: id,
      name: name,
      email: email
    };
    if (password1 !== "") data["newpassword"] = password1;

    try {
      const res = await axios.put(endpoint, data, config);

      if (res.status === 200) window.location.reload();
      else {
        createNotificationError(
          idNotificationGenerator(),
          "Something went wrong",
          "Try again later"
        )(store.dispatch);
      }
    } catch (err) {
      createNotificationError(
        idNotificationGenerator(),
        "Something went wrong",
        "Try again later"
      )(store.dispatch);
    }
  }
}
