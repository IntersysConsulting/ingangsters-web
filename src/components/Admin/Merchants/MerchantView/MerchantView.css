.merchantForm {
  width: 100%;
  max-width: 1000px;
  margin: auto;
}

.showPasswordButton {
  font-size: 30px;
  cursor: pointer;
  margin-top: 5px;
}

.showPasswordButton:hover,
.showPasswordButton.active {
  color: var(--light-blue) !important;
}

.passwordField[type="text"] {
  color: #d5d5d5;
}

.deleteAdmin {
  background-color: var(--red);
  border-color: var(--red);
  color: white;
}

.merchantForm .bottomButtonBar {
  display: flex;
  justify-content: space-around;
}

.merchantForm .bottomButtonBar > .saveButton {
  color: white;
  background-color: var(--light-blue);
  border-color: var(--light-blue);
}

.merchantForm .bottomButtonBar > .backButton {
  color: white;
  background-color: var(--dark-grey);
  border-color: var(--dark-grey);
}
