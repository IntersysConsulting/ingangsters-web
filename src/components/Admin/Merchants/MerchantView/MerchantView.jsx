import React, { useState } from "react";
import LoadingBall from "../../../UI/Loading/Loading";
import Bar from "../../../NavBars/AdminNavBar/AdminNavBar";
import Tabs from "../../AdminTabBar/AdminTabBar";
import { Row, Col, Form, Button } from "react-bootstrap";
import { FaEye } from "react-icons/fa";
import "./MerchantView.css";
import { loadAdmin, deleteAdmin, createNewAdmin } from "./Connections";
import ConfirmationModal from "../../../UI/ConfirmationModal/ConfirmationModal";
const MerchantView = ({ match }) => {
  const { id } = match.params;
  const [showPassword, setShowPassword] = useState(false);
  const [showPasswordConfirm, setShowPasswordConfirm] = useState(false);
  const [formData, setFormData] = useState({
    showDeleteButton: false,
    title: "Add Administrator",
    name: "",
    lastName: "",
    email: "",
    phone: "",
    password: "",
    passwordConf: "",
    submitAction: createNewAdmin,
    isLoading: id !== "new",
    showConfirmationModal: false,
    requirePassword: true
  });
  let content;
  const toggleConfirmationModal = () => {
    setFormData(prev => ({
      ...prev,
      showConfirmationModal: !prev.showConfirmationModal
    }));
  };
  if (formData.isLoading) {
    loadAdmin(id, setFormData);
    content = <LoadingBall />;
  } else {
    content = (
      <Form
        className="merchantForm"
        onSubmit={evt => {
          evt.preventDefault();
          formData.submitAction(evt);
          return false;
        }}
      >
        <h3>{formData.title}</h3>
        <br />
        <Row>
          <Form.Group as={Col}>
            <Form.Control
              type="text"
              name="merchantName"
              placeholder="Name"
              defaultValue={formData.name}
              required
            />
          </Form.Group>
          <Form.Group as={Col}>
            <Form.Control
              type="text"
              name="merchantLastName"
              placeholder="Last name"
              defaultValue={formData.lastName}
              required
            />
          </Form.Group>
        </Row>
        <Row>
          <Form.Group as={Col}>
            <Form.Control
              type="email"
              name="merchantEmail"
              placeholder="Email"
              defaultValue={formData.email}
              required
            />
          </Form.Group>
        </Row>
        <Row>
          <Form.Group as={Col}>
            <Form.Control
              type="phone"
              name="merchantPhone"
              placeholder="Phone (Optional)"
              defaultValue={formData.phone}
            />
          </Form.Group>
        </Row>
        <Row>
          <Form.Group as={Col} xs={9} md={10} lg={11}>
            <Form.Control
              type={showPassword ? "text" : "password"}
              name="merchantPassword"
              className="passwordField"
              placeholder="Password"
              required={formData.requirePassword}
              onChange={evt => {
                const x = evt.target.value;
                setFormData(prev => ({
                  ...prev,
                  requirePassword: x !== ""
                }));
              }}
            ></Form.Control>
          </Form.Group>
          <Col xs={3} md={2} lg={1}>
            <FaEye
              color="#000"
              className={"showPasswordButton" + (showPassword ? " active" : "")}
              onClick={() => {
                setShowPassword(!showPassword);
              }}
            />
          </Col>
        </Row>
        <Row>
          <Form.Group as={Col} xs={9} md={10} lg={11}>
            <Form.Control
              type={showPasswordConfirm ? "text" : "password"}
              className="passwordField"
              name="merchantPasswordConfirm"
              placeholder="Confirm password"
              required={formData.requirePassword}
            ></Form.Control>
          </Form.Group>
          <Col xs={3} md={2} lg={1}>
            <FaEye
              color="#000"
              className={
                "showPasswordButton" + (showPasswordConfirm ? " active" : "")
              }
              onClick={() => {
                setShowPasswordConfirm(!showPasswordConfirm);
              }}
            />
          </Col>
        </Row>
        <Row className="bottomButtonBar">
          <Button variant="primary" type="submit" className="saveButton">
            Save
          </Button>
          <Button
            className={
              "deleteAdmin" + (formData.showDeleteButton ? "" : " hide")
            }
            onClick={toggleConfirmationModal}
          >
            Delete
          </Button>
        </Row>
        <ConfirmationModal
          title="Delete admin"
          message="Are you sure you want to delete this administrator? This is a PERMANENT operation"
          affirmativeText="Confirm"
          affirmativeAction={() => {
            deleteAdmin(id);
            toggleConfirmationModal();
          }}
          negativeText="No"
          negativeAction={toggleConfirmationModal}
          closeAction={toggleConfirmationModal}
          show={formData.showConfirmationModal}
        />
      </Form>
    );
  }

  return (
    <React.Fragment>
      <Bar />
      <Tabs />
      {content}
    </React.Fragment>
  );
};

export default MerchantView;
