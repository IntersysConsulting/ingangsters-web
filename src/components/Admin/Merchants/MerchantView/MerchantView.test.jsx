import React from "react";
import { shallow } from "enzyme";
import MerchantView from "./MerchantView";

describe("Merchant View", () => {
  it('Renders "new admin" form', () => {
    const shallowRender = shallow(
      <MerchantView match={{ params: { id: "new" } }} />
    );
    expect(shallowRender).toMatchSnapshot();
  });

  it("Renders loading animation when admin ID passed", () => {
    const shallowRender = shallow(
      <MerchantView match={{ params: { id: "5da74ef4677c765948cd8d80" } }} />
    );
    expect(shallowRender.children().length).toBe(3);
  });
});
