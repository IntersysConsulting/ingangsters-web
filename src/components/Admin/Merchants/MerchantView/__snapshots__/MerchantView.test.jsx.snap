// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merchant View Renders "new admin" form 1`] = `
<Fragment>
  <Connect(NavBarComponent) />
  <Connect(AdminTabBar) />
  <Form
    className="merchantForm"
    inline={false}
    onSubmit={[Function]}
  >
    <h3>
      Add Administrator
    </h3>
    <br />
    <ForwardRef
      noGutters={false}
    >
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
      >
        <FormControl
          defaultValue=""
          name="merchantName"
          placeholder="Name"
          required={true}
          type="text"
        />
      </FormGroup>
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
      >
        <FormControl
          defaultValue=""
          name="merchantLastName"
          placeholder="Last name"
          required={true}
          type="text"
        />
      </FormGroup>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
      >
        <FormControl
          defaultValue=""
          name="merchantEmail"
          placeholder="Email"
          required={true}
          type="email"
        />
      </FormGroup>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
      >
        <FormControl
          defaultValue=""
          name="merchantPhone"
          placeholder="Phone (Optional)"
          type="phone"
        />
      </FormGroup>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
        lg={11}
        md={10}
        xs={9}
      >
        <FormControl
          className="passwordField"
          name="merchantPassword"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
      </FormGroup>
      <Col
        lg={1}
        md={2}
        xs={3}
      >
        <FaEye
          className="showPasswordButton"
          color="#000"
          onClick={[Function]}
        />
      </Col>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
        lg={11}
        md={10}
        xs={9}
      >
        <FormControl
          className="passwordField"
          name="merchantPasswordConfirm"
          placeholder="Confirm password"
          required={true}
          type="password"
        />
      </FormGroup>
      <Col
        lg={1}
        md={2}
        xs={3}
      >
        <FaEye
          className="showPasswordButton"
          color="#000"
          onClick={[Function]}
        />
      </Col>
    </ForwardRef>
    <ForwardRef
      className="bottomButtonBar"
      noGutters={false}
    >
      <Button
        active={false}
        className="saveButton"
        disabled={false}
        type="submit"
        variant="primary"
      >
        Save
      </Button>
      <Button
        active={false}
        className="deleteAdmin hide"
        disabled={false}
        onClick={[Function]}
        type="button"
        variant="primary"
      >
        Delete
      </Button>
    </ForwardRef>
    <ConfirmationModal
      affirmativeAction={[Function]}
      affirmativeText="Confirm"
      closeAction={[Function]}
      message="Are you sure you want to delete this administrator? This is a PERMANENT operation"
      negativeAction={[Function]}
      negativeText="No"
      show={false}
      title="Delete admin"
    />
  </Form>
</Fragment>
`;
